plugins/wacom-i2c/fu-wac-aes-device.c

```diff
@@ -0,0 +1,180 @@
+		0x00,				/* blkNo */
+	};
+	if (!fu_wac_device_cmd (FU_WAC_DEVICE (self),
+				cmd, sizeof(cmd), rsp, sizeof(rsp),
+				2000 * 1000, /* this takes a long time */
+				FU_WAC_DEVICE_CMD_FLAG_POLL_ON_WAITING, error)) {
+		g_prefix_error (error, "failed to send eraseall command: ");
+		return FALSE;
+	}
+	if (!fu_wac_common_rc_set_error (rsp[RTRN_RSP], error)) {
+		g_prefix_error (error, "failed to erase");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_aes_device_write_block (FuWacAesDevice *self,
+			       guint32 idx,
+			       guint32 address,
+			       const guint8 *data,
+			       guint16 datasz,
+			       GError **error)
+{
+	guint blocksz = fu_wac_device_get_block_sz (FU_WAC_DEVICE (self));
+	guint8 rsp[RSP_SIZE];
+	g_autofree guint8 *cmd = g_malloc0 (blocksz + 8);
+	cmd[0] = FU_WAC_DEVICE_BL_REPORT_ID_CMD;
+	cmd[1] = FU_WAC_DEVICE_BL_CMD_WRITE_FLASH;
+	cmd[2] = (guint8) idx;	/* echo */
+
+	/* address */
+	fu_common_write_uint32 (cmd + 3, address, G_LITTLE_ENDIAN);
+
+	/* size */
+	if (datasz != blocksz) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "block size 0x%x != 0x%x untested",
+			     datasz, (guint) blocksz);
+		return FALSE;
+	}
+	cmd[7] = datasz / 8;
+
+	/* data */
+	memcpy (cmd + 8, data, datasz);
+
+	/* write */
+	if (!fu_wac_device_cmd (FU_WAC_DEVICE (self),
+				cmd, sizeof(cmd), rsp, sizeof(rsp), 1000,
+				FU_WAC_DEVICE_CMD_FLAG_NONE, error)) {
+		g_prefix_error (error, "failed to write block %u: ", idx);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_aes_device_write_firmware (FuDevice *device, GPtrArray *chunks, GError **error)
+{
+	FuWacAesDevice *self = FU_WAC_AES_DEVICE (device);
+
+	/* erase */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_ERASE);
+	if (!fu_wac_aes_device_erase_all (self, error))
+		return FALSE;
+
+	/* write */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_wac_aes_device_write_block (self,
+						    chk->idx,
+						    chk->address,
+						    chk->data,
+						    chk->data_sz,
+						    error))
+			return FALSE;
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+	return TRUE;
+}
+
+static void
+fu_wac_aes_device_init (FuWacAesDevice *self)
+{
+}
+
+static void
+fu_wac_aes_device_class_init (FuWacAesDeviceClass *klass)
+{
+	FuDeviceClass *klass_device = FU_DEVICE_CLASS (klass);
+	FuWacDeviceClass *klass_wac_device = FU_WAC_DEVICE_CLASS (klass);
+	klass_device->setup = fu_wac_aes_device_setup;
+	klass_wac_device->write_firmware = fu_wac_aes_device_write_firmware;
+}
+
+FuWacAesDevice *
+fu_wac_aes_device_new (FuUdevDevice *device)
+{
+	FuWacAesDevice *self = g_object_new (FU_TYPE_WAC_AES_DEVICE,
+					     "i2c-addr", 0x0a,
+					     "flash-block-size", 128,
+					     "flash-base-addr", 0x8000,
+					     "flash-size", 0x24000,
+					     NULL);
```

Comment:
These feel like they should be coming from quirks

Reply:
Agreed, I'll fix this up.